dj/models/node.py

```diff
@@ -152,6 +152,44 @@ class NodeMissingParents(SQLModel, table=True):  # type: ignore
     )
 
 
+class AvailabilityStateBase(SQLModel):
+    """
+    An availability state base
+    """
+
+    catalog: Optional[str] = None
+    schema_: Optional[str] = Field(default=None)
+    table: str
+    valid_through_ts: int
+    max_partition: str
```

Comment:
@samredai if you want to take a shortcut for now and say that all the partition keys will be serialized in the correct order to a single string, that's ok, but we should at least make that comment in here. Ideally we have a list of values in here, ordered according to the partition key list, e.g. dateint, hourint, etc.

Reply:
Thanks @agorajek, I'm definitely leaning on you here in order to understand the exact information/functionality that needs to exist here. Since a table can be partitioned by multiple fields with a hierarchy, it seems like we may need a separate attribute specifically to determine that hierarchy, right? For example, let's take a table that has separate `year`, `month`, `day`, `country` partitions. We would need to know something like:
```yaml
partitions: ["year", "month", "day", "country"]
max_partition: "2023,01,26,AF"
min_partition: "2020,01,01,ZW"
```
We could also go all the way and include type information with a `datepart` type so we know how to properly compare the incoming `AvailabilityState` events to the existing `AvailabilityState` and do the merge properly. So maybe...
```yaml
partitions: [
    {"name": "year", "type": "datepart", "format": "yyyy"},
    {"name": "month", "type": "datepart", "format": "mm"},
    {"name": "day", "type": "datepart", "format": "dd"},
    {"name": "country", "type": "str"},
]
max_partition: 2023-01-26T00:00:00.000000  # datetime object
min_partition: 2020-01-01T00:00:00.000000  # datetime object
```
Then we'd know the last element isn't really important in determining the date range of the available data. Am I at least getting warmer? :)

Reply:
@samredai the partition spec in the form of an ordered list of field names should simply be part of the node attributes. I don't think we need to carry it inside the AvailabilityState. What we need to carry here is the values for all or some of the first partition keys (according to the order from the spec). This way we can keep these values as a dict (or a list) and serialize to a string for comparison whenever we need to compare two states.

One thing that I just realized is that when we update nodes (specifically their schemas or partition keys) we will need to invalidate their AvailabilityState (most likely) and count on the Reflection service to pick up the slack and let us know the new AvailabilityState on the changed nodes.

Reply:
Got it, makes sense!

Reply:
Updated the PR to use `List[str]` for the partition values. Surprisingly, python already sorts them the way you'd expect so I didn't have to change any of the merge logic. I added a comment there though about the need for considering the partition types and doing a cast before comparing.
```py
>>> max(["2022", "01", "US"], ["2019", "01", "ZW"])
['2022', '01', 'US']
>>> max(["2022", "01", "01"], ["2022", "02", "01"])
['2022', '02', '01']
```